crates/apollo-compiler/src/ast/serialize.rs

```diff
@@ -831,6 +842,89 @@ fn serialize_string_value(state: &mut State, mut str: &str) -> fmt::Result {
+    const TRIPLE_QUOTE: &str = "\"\"\"";
+    const ESCAPED_TRIPLE_QUOTE: &str = "\\\"\"\"";
+    const _: () = assert!(TRIPLE_QUOTE.len() == 3);
+    const _: () = assert!(ESCAPED_TRIPLE_QUOTE.len() == 4);
+
+    state.write(TRIPLE_QUOTE)?;
+    // `can_be_block_string` excludes \r, so the only remaining line terminator is \n
+    for mut line in str.split('\n') {
+        if line.is_empty() {
+            // Skip indentation which would be trailing whitespace
+            state.write("\n")?;
+            continue;
+        }
+        state.require_new_line()?;
+        loop {
+            if let Some((before, after)) = line.split_once(TRIPLE_QUOTE) {
+                state.write(before)?;
+                state.write(ESCAPED_TRIPLE_QUOTE)?;
+                line = after;
+            } else {
+                state.write(line)?;
+                break;
+            }
+        }
+    }
+    state.require_new_line()?;
+    state.write(TRIPLE_QUOTE)
+}
+
+fn prefer_block_string(value: &str) -> bool {
+    // If a value is multi-line, it likely looks nicer as a block string
+    value.contains('\n')
+}
+
+/// Is it possible to create a serialization that, when fed through
+/// [BlockStringValue](https://spec.graphql.org/October2021/#BlockStringValue()),
+/// returns exactly `value`?
+fn can_be_block_string(value: &str) -> bool {
+    // `BlockStringValue` splits its inputs at any `LineTerminator` (\n, \r\n, or \r)
+    // and eventually joins lines but always with \n. So its output can never contain \r
+    if value.contains('\r') {
+        return false;
+    }
+
+    /// <https://spec.graphql.org/October2021/#WhiteSpace>
+    fn trim_start_graphql_whitespace(value: &str) -> &str {
+        value.trim_start_matches([' ', '\t'])
+    }
+
+    // With the above, \n is the only remaining LineTerminator
+    let mut lines = value.split('\n');
+    if lines
+        .next()
+        .is_some_and(|first| trim_start_graphql_whitespace(first).is_empty())
+        || lines
+            .next_back()
+            .is_some_and(|last| trim_start_graphql_whitespace(last).is_empty())
+    {
+        // Leading or trailing whitespace-only line would be trimmed by `BlockStringValue`
+        return false;
+    }
+
+    let common_indent = {
+        let mut lines = value.split('\n');
+        // Skip the first line, the one following """
+        lines.next();
+
+        let each_line_indent_utf8_len = lines.filter_map(|line| {
+            let after_indent = trim_start_graphql_whitespace(line);
+            if !after_indent.is_empty() {
+                Some(line.len() - after_indent.len())
+            } else {
+                None // skip whitespace-only lines
+            }
+        });
+        each_line_indent_utf8_len.min().unwrap_or(0)
+    };
+    // If there is common indent `BlockStringValue` would remove it
+    // and incorrectly round-trip to a different value.
+    common_indent == 0
+}
+
```

Comment:
It is worth noting that each time the function `value.contains` is called, it iterates through the entire string. The same goes for each call of `value.split`. 

However, graphql-js optimizes this process by performing a single iteration through the string. This optimization is particularly useful for long descriptions. 
https://github.com/graphql/graphql-js/blob/0b7590f0a2b65e6210da2e49be0d8e6c27781af2/src/language/blockString.ts#L54.